Zend/zend_ast.c

```diff
@@ -420,6 +420,25 @@ ZEND_API zend_ast *zend_ast_create_list(uint32_t init_children, zend_ast_kind ki
 }
 #endif
 
+zend_ast *zend_ast_create_concat_op(zend_ast *op0, zend_ast *op1) {
+	zend_ast *result = zend_ast_create_binary_op(ZEND_CONCAT, op0, op1);
```

Comment:
Can this allocation be avoided if the merging is successful? It also looks like leaks in that case but maybe I'm missing something.

```c
if (zend_binary_op_produces_error(ZEND_CONCAT, zv0, zv1)) {
    goto failure;
}
if (concat_function(zv0, zv0, zv1) == FAILURE) {
    goto failure;
}
zval_ptr_dtor_nogc(zv1);
return zend_ast_create_zval(zv0);
failure:
return zend_ast_create_binary_op(ZEND_CONCAT, op0, op1);
```

Reply:
This was left over from before refactoring this to call concat_function directly and can be cleaned up safely.

Reply:
```c
static inline void *zend_ast_alloc(size_t size) {
	return zend_arena_alloc(&CG(ast_arena), size);
}
```

>It also looks like leaks in that case but maybe I'm missing something.

Zend/zend_compile.c calls zend_arena_alloc and zend_arena_free to free all of the memory used by the arena at once, so the "leak" isn't actually a leak after compilation finishes, it's just slightly inefficient.